src/QsCompiler/CompilationManager/DataStructures.cs

```diff
@@ -41,69 +50,357 @@ internal class CodeLine
+                        {
+                            var next = text.Substring(index + 1).IndexOfAny(new[] { '"', '{' });
+                            index = next < 0 ? next : index + 1 + next;
+                        }
+                        break;
+                    case StringContext.OpenString:
+                    case StringContext.OpenStringInOpenInterpolatedArgument:
+                        // Find the next " not preceded by \
+                        index = text.IndexOf('"');
+                        while (index > 0 && text[index - 1] == '\\')
+                        {
+                            var next = text.Substring(index + 1).IndexOf('"');
+                            index = next < 0 ? next : index + 1 + next;
+                        }
+                        break;
+                }
+
+                if (index < 0 || (commentIndex > -1 && commentIndex < index))
+                {
+                    break;
+                }
+
+                var inputDelimiter = text[index].ToString();
+                // If the input is a " preceded by a $ than the input needs to be updated to $"
+                if (index > 0 && inputDelimiter == "\"" && text[index - 1] == '$')
+                {
+                    inputDelimiter = "$\"";
+                }
+
+                var pos = index + stringLength - text.Length;
+                bool validDelimiter;
+                (stringContext, validDelimiter) = MoveToNextState(stringContext, inputDelimiter);
+                if (!validDelimiter)
+                {
+                    // If the erroneous delimiter is $", mark the $ as the error and treat the " as a normal string delimiter
+                    if (inputDelimiter == "$\"")
+                    {
+                        errorDelimiterBuilder.Add(pos - 1); // Backup to the $ character
+                        delimiterBuilder.Add(pos);
+                    }
+                    else
+                    {
+                        errorDelimiterBuilder.Add(pos);
+                    }
+                }
+                else
+                {
+                    delimiterBuilder.Add(pos);
+                }
+                text = text.Substring(index + 1);
+            }
+
+            // commentIndex is only nonzero if we found \\ and we were not in a string,
+            // so it should be the genuine start of a comment
+            if (commentIndex > -1)
+            {
+                commentIndex += stringLength - text.Length;
+            }
+
+            if (stringContext >= StringContext.OpenString)
+            {
+                delimiterBuilder.Add(stringLength);
+            }
+
+            errorDelimiters = errorDelimiterBuilder.ToImmutable();
+            return delimiterBuilder.ToImmutable();
         }
 
-        internal CodeLine SetCommentIndex(int newCommentIndex)
+        private static (StringContext, bool) MoveToNextState(StringContext curr, string input)
```

Comment:
Would be helpful to add a comment about what bool means in return tuple. Alternatively create a named type with the named field.

Reply:
Thanks! I've given names to the tuple fields to make it clearer what the fields are for.